In [1]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error

In [12]:
df = pd.read_csv('../data/processed/dept35_22.csv')

In [13]:
df.head()

Sku Number  Year Number  Month Number  Store Number  Total Sales  \
0      163320           22             1             1        38.97   
1      163320           22             1             5        25.98   
2      163320           22             1             6        38.97   
3      163320           22             1             7        25.98   
4      163320           22             1             8        25.98   

   Total Units Sold  Total OH Inventory Cost  Total OH Inventory Unit  
0               3.0                    22.06                      3.0  
1               2.0                    27.03                      4.0  
2               3.0                    14.79                      2.0  
3               2.0                    34.20                      5.0  
4               2.0                    41.27                      6.0

In [14]:
%%time

X = df.drop('Total Units Sold', axis=1)
y = df['Total Units Sold']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LinearRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)

print(f'MLR RMSE: {round(np.sqrt(mean_squared_error(y_test, predictions)), 3)}')
print(f'MLR R^2: {round(r2_score(y_test, predictions), 3)} \n')

MLR RMSE: 1.375
MLR R^2: 0.698 

CPU times: user 285 ms, sys: 60.1 ms, total: 345 ms
Wall time: 190 ms


In [49]:
df.sample(5)

Sku Number  Year Number  Month Number  Store Number  Total Sales  \
110007      543033           22             1            20         0.00   
532952      632290           22             8            23         0.00   
16109       356533           22             3            50        13.97   
548795      633775           22             8             1       189.81   
152929      573764           22             7            46         9.98   

        Total Units Sold  Total OH Inventory Cost  Total OH Inventory Unit  
110007               0.0                    11.57                      8.0  
532952               0.0                    26.21                      6.0  
16109                3.0                    52.26                     20.0  
548795              19.0                    77.64                     13.0  
152929               2.0                    56.02                     17.0

In [50]:
jan_X = df[(df['Month Number'] == 8) & (df['Sku Number'] == 633775)].drop('Total Units Sold', axis=1)
jan_y = df[(df['Month Number'] == 8) & (df['Sku Number'] == 633775)]['Total Units Sold']

In [51]:
df[(df['Month Number'] == 8) & (df['Sku Number'] == 633775)]

Sku Number  Year Number  Month Number  Store Number  Total Sales  \
548795      633775           22             8             1       189.81   
548796      633775           22             8             5        99.90   
548797      633775           22             8             6        79.92   
548798      633775           22             8             7        59.94   
548799      633775           22             8             8        19.98   
548800      633775           22             8            11        49.95   
548801      633775           22             8            14        19.98   
548802      633775           22             8            17       329.67   
548803      633775           22             8            20        69.93   
548804      633775           22             8            23       119.88   
548805      633775           22             8            24       149.85   
548806      633775           22             8            25       139.86   
548807      633775           22             8            26       159.84   
548808      633775           22             8            27       429.57   
548809      633775           22             8            28       239.76   
548810      633775           22             8            29       209.79   
548811      633775           22             8            31       219.78   
548812      633775           22             8            32       209.79   
548813      633775           22             8            33       147.85   
548814      633775           22             8            36       129.87   
548815      633775           22             8            37       129.87   
548816      633775           22             8            39       819.18   
548817      633775           22             8            40         0.00   
548818      633775           22             8            41         0.00   
548819      633775           22             8            43       109.89   
548820      633775           22             8            44         0.00   
548821      633775           22             8            46        59.94   
548822      633775           22             8            47        67.93   
548823      633775           22             8            48        79.92   
548824      633775           22             8            50        39.96   
548825      633775           22             8            51         0.00   
548826      633775           22             8            52       159.84   
548827      633775           22             8            53       139.86   
548828      633775           22             8            54       169.83   
548829      633775           22             8            55        99.90   
548830      633775           22             8            56        39.96   
548831      633775           22             8            57        29.97   
548832      633775           22             8            58        39.96   

        Total Units Sold  Total OH Inventory Cost  Total OH Inventory Unit  
548795              19.0                    77.64                     13.0  
548796              10.0                    81.80                     14.0  
548797               8.0                    17.55                      3.0  
548798               6.0                    54.28                      9.0  
548799               2.0                    12.66                      2.0  
548800               5.0                    31.87                      5.0  
548801               2.0                    52.23                      8.0  
548802              33.0                   481.92                     77.0  
548803               7.0                    53.85                      9.0  
548804              12.0                    53.55                      9.0  
548805              15.0                    30.54                      5.0  
548806              14.0                    95.13                     16.0  
548807              16.0                  

In [52]:
print('Predicted vs. Actual: Difference/Error')
print('-' * 40)
for x in range(len(jan_X.index)):
    prediction = np.round(model.predict(pd.DataFrame([jan_X.iloc[x]])))
    print(prediction, '\t\t', jan_y.iloc[x], '\t\t', prediction - jan_y.iloc[x])
    # print(prediction+1, '\t\t', jan_y.iloc[x], '\t\t', prediction+1 - jan_y.iloc[x])

Predicted vs. Actual: Difference/Error
----------------------------------------
[16.] 		 19.0 		 [-3.]
[8.] 		 10.0 		 [-2.]
[7.] 		 8.0 		 [-1.]
[5.] 		 6.0 		 [-1.]
[2.] 		 2.0 		 [0.]
[4.] 		 5.0 		 [-1.]
[2.] 		 2.0 		 [0.]
[27.] 		 33.0 		 [-6.]
[6.] 		 7.0 		 [-1.]
[10.] 		 12.0 		 [-2.]
[13.] 		 15.0 		 [-2.]
[12.] 		 14.0 		 [-2.]
[13.] 		 16.0 		 [-3.]
[36.] 		 43.0 		 [-7.]
[20.] 		 24.0 		 [-4.]
[18.] 		 21.0 		 [-3.]
[18.] 		 22.0 		 [-4.]
[18.] 		 21.0 		 [-3.]
[12.] 		 15.0 		 [-3.]
[11.] 		 13.0 		 [-2.]
[11.] 		 13.0 		 [-2.]
[69.] 		 82.0 		 [-13.]
[0.] 		 0.0 		 [0.]
[-0.] 		 0.0 		 [-0.]
[9.] 		 11.0 		 [-2.]
[-0.] 		 0.0 		 [-0.]
[5.] 		 6.0 		 [-1.]
[6.] 		 7.0 		 [-1.]
[7.] 		 8.0 		 [-1.]
[3.] 		 4.0 		 [-1.]
[0.] 		 0.0 		 [0.]
[13.] 		 16.0 		 [-3.]
[12.] 		 14.0 		 [-2.]
[14.] 		 17.0 		 [-3.]
[8.] 		 10.0 		 [-2.]
[3.] 		 4.0 		 [-1.]
[3.] 		 3.0 		 [0.]
[3.] 		 4.0 		 [-1.]


SO, MLR is ~35x faster than XGB, and honestly not even that inaccuracte, since we aren't going for exacts. XGB is ~26% more accuracte, but again MLR is still within a +- 1 range, which is incredible for this tool

#### ask [redacted] if overstocking by a unit or two could be detrimental, i.e. how MLR will +1 most of the time

### ALSO USE THIS SAMPLE | Month: 8, SKU: 633775 | hella units sold